packages/gatsby-plugin-fastify/src/plugins/headers.ts

```diff
@@ -0,0 +1,46 @@
+import fp from "fastify-plugin";
+
+import { SECURITY_HEADERS } from "../utils/constants";
+import { removeQueryParmsFromUrl } from "../utils/routes";
+
+import type { Headers } from "../utils/headers";
+import type { FastifyPluginAsync, FastifyReply } from "fastify";
+
+export const handleHeaders: FastifyPluginAsync<{
+  headers: Headers;
+  useDefaultSecurity: boolean;
+}> = fp(async (fastify, { headers, useDefaultSecurity }) => {
+  fastify.addHook("onSend", async function (request, reply: FastifyReply) {
+    let pageHeaders = {};
+    const useServerHeaders = ["FUNCTION", "SSR", "PROXY"].includes(reply.mode);
+    const serverHeaders = useServerHeaders ? reply.getHeaders() : {};
+
+    // apply default security headers here instead of header-builder to
+    // prevent duplication on every page/file entry in the postbuild config
+    if (useDefaultSecurity) pageHeaders = { ...SECURITY_HEADERS["/**"] };
+
+    // get the url without query params
+    const workingURL = removeQueryParmsFromUrl(request.url);
+    const urlHeaders = headers[workingURL];
+    if (urlHeaders) pageHeaders = { ...pageHeaders, ...urlHeaders };
+
+    // handle special path type routes
+    if (["CSR", "SSR", "REDIRECT", "FUNCTION", "PROXY", "404", "500"].includes(reply.mode)) {
```

Comment:
So I thought about something after I went to bed. If you look at the `x-gatsby-server` headers that get appended by every module some routes get multiple modules appended. e.g. 404 and 500 routes hit those handlers but because they use the static file server the static handler is hit last. Have you confirmed you will ever actually get a 400 or 500? Some CSR might be similar.

Reply:
Yes, `404`, `500`, and `CSR`, set `reply.mode` and `reply.path` at the same time they append the module header, ie: 

```js
    reply.appendModuleHeader("404");
    reply.mode = "404";
    reply.path = "/404/";
```

So they look for the configured headers for their path, ie: `/404/`, even though the actual URL might be something like: `/some/non-existing/page`. That being said, the user could still configure `customHeaders` for `/some/non-existing/page` with a matching pattern entry if they needed to because `urlHeaders` are also added to `pageHeaders` alongside these `pathHeaders`.

One thing I have on my to do list to check is that I probably should be adding `pathHeaders` before `urlHeaders` because `pathHeaders` would generally just contain the defaults for those modules and `urlHeaders` would contain `customHeaders` that should be able to overwrite the defaults. I'll check up on this and add a test case.

Reply:
Not sure if you undestood what I was saying `x-gatsby-server` can return something like `SSR; 404; Static;`. Cause the SSR module tried to route the url, but then 404 ed, which serves the 404 file via the Static handler. If each module is overwriting the last, your `reply.mode` would be `static` not `404` when it arrives at the headers plugin.

Reply:
Thinking on this a little more, I think your concern is that the `static` module will overwrite the `reply.mode`. That is not the case because the `reply.mode` decorator is only set on modules that use dynamic routes. Static does not set it. There are two reasons for this, one is because we only need it for dynamic routes, additionally, although it could be useful to know the `mode` even for static/other routes there is some background why I didn't/couldn't set it, at least for static, [in this comment](https://github.com/gatsby-uc/plugins/pull/392#discussion_r1136091588):

>In the first refactor I did go down the path of tagging routes with their module by adding a decorator reply.module which I was going to then use in the headers plugin to set the FG_MODULE_HEADER based on the module used as well. I had this working for all modules except static.ts because it is globally scoped, since the fastify-static plugin doesn't have any hooks to run when it's actually serving a file the Static module was added to all replies, not just the ones actually served by static.